modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -422,65 +422,133 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
 
     @Override
     public Hit getHitInfo(float x, float y, String text, int textRunStart, int curRunStart) {
+        boolean leading = false;
+        boolean isMirrored = isMirrored(); // Node orientation is RTL
+        boolean isMultiRunText = false;
         int charIndex = -1;
         int insertionIndex = -1;
-        boolean leading = false;
         int relIndex = 0;
+        int LTRIndex = 0;
         int textWidthPrevLine = 0;
+        float xHitPos = x;
 
         ensureLayout();
         int lineIndex = getLineIndex(y, text, curRunStart);
         if (lineIndex >= getLineCount()) {
             charIndex = getCharCount();
             insertionIndex = charIndex + 1;
         } else {
-            if (isMirrored()) {
+            boolean isTextFlow = textRunStart == -1 && curRunStart == -1;
+            if (isMirrored && (isTextFlow || spans == null)) {
                 x = getMirroringWidth() - x;
             }
             TextLine line = lines[lineIndex];
             TextRun[] runs = line.getRuns();
             RectBounds bounds = line.getBounds();
             TextRun run = null;
-            x -= bounds.getMinX();
             //TODO binary search
             if (text == null || spans == null) {
-                for (int i = 0; i < runs.length; i++) {
-                    run = runs[i];
-                    if (x < run.getWidth()) break;
-                    if (i + 1 < runs.length) {
-                        if (runs[i + 1].isLinebreak()) break;
-                        x -= run.getWidth();
+                // To calculate Text and TextFlow hit info
+                if (isMirrored) {
+                    int runIndex = -1;
+                    for (int i = runs.length - 1; i >=0; i--) {
+                        run = runs[i];
+                        if (x < run.getWidth() && (isTextFlow || (run.getStart() == curRunStart))) {
+                            runIndex = i;
+                            break;
+                        }
+                        if (i - 1 >= 0) {
+                            if (runs[i - 1].isLinebreak()) break;
+                            x -= run.getWidth();
+                        }
+                    }
+                    for (int i = 0; i < runIndex; i++) {
+                        xHitPos -= runs[i].getWidth();
+                    }
+                    xHitPos -= bounds.getMinX();
+                } else {
+                    for (int i = 0; i < runs.length; i++) {
+                        run = runs[i];
+                        if (x < run.getWidth()) break;
+                        if (i + 1 < runs.length) {
+                            if (runs[i + 1].isLinebreak()) break;
+                            x -= run.getWidth();
+                        }
                     }
                 }
             } else {
+                // To calculate hit info of Text embedded in TextFlow
```

Comment:
the comments on lines 480 and 451 are a bit confusing: both refer to Text.  could you please clarify?